Add a blog post about using a custom query (#251)
infinum/eightshift-docs: website/blog/2023-08-03-adding-custom-query.md

---

> title: Adding a custom query

Reviewer: ```suggestion
title: Working with custom queries
```

Author: Yes, absolutely! 😄 Coming up with a good title on this one was killing me 😅

description: Explains the process of registering a new service class, adding a custom query and using it in a block.

> slug: adding-custom-query

Reviewer: ```suggestion
slug: working-with-custom-queries
```
(if you decide to accept the above one)

authors: obradovic

date: 2023-08-03
tags: [eightshift, boilerplate, service, class, query]
hide_table_of_contents: false
---

> Eightshift Dev Kit offers some blocks with query logic out-of-the-box, but what is the best approach when you need to add a custom query to a block you’ve been working on?

Reviewer: ```suggestion
Eightshift DevKit offers some blocks with query logic out of the box, but what is the best approach when you need to add a custom query to a block you’ve been working on?
```

<!--truncate-->

> ## Already existing blocks

Reviewer: ```suggestion
## Usage in built-in blocks
```

```php
\add_filter(self::GET_POSTS_BY_CATEGORY, [$this, 'getPostsByCategory'], 10, 3);
```

## Using the new filter

:::note
Even if you register a filter for a method that doesn’t accept any arguments, when calling `apply_filters`, you must pass at least 1 parameter. In those cases, simply add `null` as a parameter.
:::

Here are some examples how you can use the filter:

```php
// Get 3 latest posts, regardless of category.
$allLatestPosts = apply_filters(CustomQuery::GET_POSTS_BY_CATEGORY, null);

// Get 10 latest posts from the News category. News category ID is 2.
$latestNews = apply_filters(CustomQuery::GET_POSTS_BY_CATEGORY, 2, 10);

// Get another page of News category.
$pagedNews = apply_filters(CustomQuery::GET_POSTS_BY_CATEGORY, 2, 10, $currentPage);
```

The above example shows multiple use cases. The first two examples could be used in a simple block that displays only the selected number of the latest posts. The last example may be used in a REST route for a load more functionality or in a block with classic pagination.

> Now you can do a regular query loop in your block to display the posts:

Reviewer: Even though it's there, maybe we should mention that `wp_reset_postdata()` shouldn't be forgotten.

Author: Yep, I agree. I'll update it 🙂

description: Explains the process of registering a new service class, adding a custom query and using it in a block.
slug: working-with-custom-queries
authors: obradovic

date: 2023-08-01
tags: [eightshift, boilerplate, service, class, query]
hide_table_of_contents: false
---

Eightshift DevKit offers some blocks with query logic out of the box, but what is the best approach when you need to add a custom query to a block you’ve been working on?
<!--truncate-->

## Usage in built-in blocks

One of the Eightshift blocks that already uses `WP_Query` is the ***Featured Content*** block. In that block, you can see the query logic is inside the block. However, there is a much better way to do it. The reason it was done like this was to simplify this block and to have an already functioning block available with one WP-CLI command.

A much better approach would be to separate the query logic from the block. Other than following the MVC architecture more closely, this will also make the query logic more reusable. To do this, we’re gonna create a service class.

## What are Service classes?

*Put simply, a Service is any PHP object that performs some sort of "global" task. - Symfony docs*

> If you take a look at the `ServiceInterface` interface, you’ll notice it only  contains the `register()` method. This method holds action and filter hooks for that class. Other than hooking into existing actions and filters, this method can be used to register our own filters which can be used in blocks or other classes.

Reviewer 2: ```suggestion
If you take a look at the `ServiceInterface` interface, you’ll notice it only contains the `register()` method. This method holds action and filter hooks for that class. Other than hooking into existing actions and filters, this method can be used to register our own filters which can be used in blocks or other classes.
```

> Basically, whenever you need to hook into actions or filters, you use a Service class for that.

Reviewer 2: ```suggestion
Basically, whenever you need to hook into actions or filters, you should use a Service class for that.
```
